Data race in `SnapshotableHashTable`

```
E           Exception: Sanitizer assert found for instance ==================
E           WARNING: ThreadSanitizer: data race (pid=1)
E             Read of size 1 at 0x7b6000004100 by thread T593 (mutexes: write M0, write M1):
E               #0 DB::SnapshotableHashTable<DB::KeeperStorage::Node>::updateValue(StringRef, std::__1::function<void (DB::KeeperStorage::Node&)>) build_docker/./src/Coordination/SnapshotableHashTable.h:327:13 (clickhouse+0x19e462a5) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #1 Coordination::Error DB::KeeperStorage::commit(long)::$_0::operator()<DB::KeeperStorage::UpdateNodeDelta>(DB::KeeperStorage::UpdateNodeDelta&) const build_docker/./src/Coordination/KeeperStorage.cpp:837:51 (clickhouse+0x19eb1d41) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #2 decltype(std::declval<DB::KeeperStorage::commit(long)::$_0>()(std::declval<DB::KeeperStorage::UpdateNodeDelta&>())) std::__1::__invoke[abi:v15000]<DB::KeeperStorage::commit(long)::$_0, DB::KeeperStorage::UpdateNodeDelta&>(DB::KeeperStorage::commit(long)::$_0&&, DB::KeeperStorage::UpdateNodeDelta&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x19eb1d41)
E               #3 decltype(auto) std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>::operator()[abi:v15000]<std::__1::__variant_detail::__alt<2ul, DB::KeeperStorage::UpdateNodeDelta>&>(std::__1::__variant_detail::__alt<2ul, DB::KeeperStorage::UpdateNodeDelta>&) const build_docker/./contrib/llvm-project/libcxx/include/variant:668:14 (clickhouse+0x19eb1d41)
E               #4 decltype(std::declval<std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>>()(std::declval<std::__1::__variant_detail::__alt<2ul, DB::KeeperStorage::UpdateNodeDelta>&>())) std::__1::__invoke[abi:v15000]<std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>, std::__1::__variant_detail::__alt<2ul, DB::KeeperStorage::UpdateNodeDelta>&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>&&, std::__1::__variant_detail::__alt<2ul, DB::KeeperStorage::UpdateNodeDelta>&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x19eb1d41)
E               #5 decltype(auto) std::__1::__variant_detail::__visitation::__base::__dispatcher<2ul>::__dispatch[abi:v15000]<std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>&&, std::__1::__variant_detail::__base<(std::__1::__variant_detail::_Trait)1, DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&) build_docker/./contrib/llvm-project/libcxx/include/variant:548:16 (clickhouse+0x19eb1d41)
E               #6 decltype(auto) std::__1::__variant_detail::__visitation::__base::__visit_alt[abi:v15000]<std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>, std::__1::__variant_detail::__impl<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>&&, std::__1::__variant_detail::__impl<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&) build_docker/./contrib/llvm-project/libcxx/include/variant:511:12 (clickhouse+0x19ed040f) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #7 decltype(auto) std::__1::__variant_detail::__visitation::__variant::__visit_alt[abi:v15000]<std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>, std::__1::variant<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&>(std::__1::__variant_detail::__visitation::__variant::__value_visitor<DB::KeeperStorage::commit(long)::$_0>&&, std::__1::variant<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&) build_docker/./contrib/llvm-project/libcxx/include/variant:618:12 (clickhouse+0x19ed040f)
E               #8 decltype(auto) std::__1::__variant_detail::__visitation::__variant::__visit_value[abi:v15000]<DB::KeeperStorage::commit(long)::$_0, std::__1::variant<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&>(DB::KeeperStorage::commit(long)::$_0&&, std::__1::variant<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&) build_docker/./contrib/llvm-project/libcxx/include/variant:637:12 (clickhouse+0x19ed040f)
E               #9 decltype(auto) std::__1::visit[abi:v15000]<DB::KeeperStorage::commit(long)::$_0, std::__1::variant<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&, void>(DB::KeeperStorage::commit(long)::$_0&&, std::__1::variant<DB::KeeperStorage::CreateNodeDelta, DB::KeeperStorage::RemoveNodeDelta, DB::KeeperStorage::UpdateNodeDelta, DB::KeeperStorage::SetACLDelta, DB::KeeperStorage::AddAuthDelta, DB::KeeperStorage::ErrorDelta, DB::KeeperStorage::SubDeltaEnd, DB::KeeperStorage::FailedMultiDelta>&) build_docker/./contrib/llvm-project/libcxx/include/variant:1713:10 (clickhouse+0x19ed040f)
E               #10 DB::KeeperStorage::commit(long) build_docker/./src/Coordination/KeeperStorage.cpp:813:23 (clickhouse+0x19ed040f)
E               #11 DB::KeeperStorageRemoveRequestProcessor::process(DB::KeeperStorage&, long) const build_docker/./src/Coordination/KeeperStorage.cpp:1401:34 (clickhouse+0x19ed040f)
E               #12 DB::KeeperStorage::processRequest(std::__1::shared_ptr<Coordination::ZooKeeperRequest> const&, long, std::__1::optional<long>, bool, bool) build_docker/./src/Coordination/KeeperStorage.cpp:2476:43 (clickhouse+0x19eab835) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #13 DB::KeeperStateMachine::commit(unsigned long, nuraft::buffer&) build_docker/./src/Coordination/KeeperStateMachine.cpp:434:28 (clickhouse+0x19e7c27c) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #14 nuraft::state_machine::commit_ext(nuraft::state_machine::ext_op_params const&) build_docker/./contrib/NuRaft/include/libnuraft/state_machine.hxx:78:16 (clickhouse+0x19e881d2) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #15 nuraft::raft_server::commit_app_log(unsigned long, std::__1::shared_ptr<nuraft::log_entry>&, bool, bool) build_docker/./contrib/NuRaft/src/handle_commit.cxx:335:33 (clickhouse+0x1ce265d6) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #16 nuraft::raft_server::commit_in_bg_exec(unsigned long, bool) build_docker/./contrib/NuRaft/src/handle_commit.cxx:265:13 (clickhouse+0x1ce2582f) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #17 nuraft::raft_server::commit_in_bg() build_docker/./contrib/NuRaft/src/handle_commit.cxx:158:9 (clickhouse+0x1ce24d1c) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #18 DB::KeeperServer::KeeperRaftServer::commit_in_bg() build_docker/./src/Coordination/KeeperServer.cpp:198:30 (clickhouse+0x19df00a9) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #19 decltype(*std::declval<nuraft::raft_server*&>().*std::declval<void (nuraft::raft_server::*&)()>()()) std::__1::__invoke[abi:v15000]<void (nuraft::raft_server::*&)(), nuraft::raft_server*&, void>(void (nuraft::raft_server::*&)(), nuraft::raft_server*&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:359:23 (clickhouse+0x1ce1dcdb) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #20 std::__1::__bind_return<void (nuraft::raft_server::*)(), std::__1::tuple<nuraft::raft_server*>, std::__1::tuple<>, __is_valid_bind_return<void (nuraft::raft_server::*)(), std::__1::tuple<nuraft::raft_server*>, std::__1::tuple<>>::value>::type std::__1::__apply_functor[abi:v15000]<void (nuraft::raft_server::*)(), std::__1::tuple<nuraft::raft_server*>, 0ul, std::__1::tuple<>>(void (nuraft::raft_server::*&)(), std::__1::tuple<nuraft::raft_server*>&, std::__1::__tuple_indices<0ul>, std::__1::tuple<>&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/bind.h:263:12 (clickhouse+0x1ce1dcdb)
E               #21 std::__1::__bind_return<void (nuraft::raft_server::*)(), std::__1::tuple<nuraft::raft_server*>, std::__1::tuple<>, __is_valid_bind_return<void (nuraft::raft_server::*)(), std::__1::tuple<nuraft::raft_server*>, std::__1::tuple<>>::value>::type std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>::operator()[abi:v15000]<>() build_docker/./contrib/llvm-project/libcxx/include/__functional/bind.h:295:20 (clickhouse+0x1ce1dcdb)
E               #22 decltype(std::declval<std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>()()) std::__1::__invoke[abi:v15000]<std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>(std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x1ce1dcdb)
E               #23 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0x1ce1dcdb)
E               #24 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, std::__1::__bind<void (nuraft::raft_server::*)(), nuraft::raft_server*>>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0x1ce1dcdb)
E           
E             Previous write of size 1 at 0x7b6000004100 by thread T571:
E               #0 DB::SnapshotableHashTable<DB::KeeperStorage::Node>::disableSnapshotMode() build_docker/./src/Coordination/SnapshotableHashTable.h:410:23 (clickhouse+0x19e3d9ed) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #1 DB::KeeperStorage::disableSnapshotMode() build_docker/./src/Coordination/KeeperStorage.h:529:44 (clickhouse+0x19e3d9ed)
E               #2 DB::KeeperStorageSnapshot::~KeeperStorageSnapshot() build_docker/./src/Coordination/KeeperSnapshotManager.cpp:563:14 (clickhouse+0x19e3d9ed)
E               #3 void std::__1::__destroy_at[abi:v15000]<DB::KeeperStorageSnapshot, 0>(DB::KeeperStorageSnapshot*) build_docker/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:63:13 (clickhouse+0x19e8bf1d) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #4 void std::__1::destroy_at[abi:v15000]<DB::KeeperStorageSnapshot, 0>(DB::KeeperStorageSnapshot*) build_docker/./contrib/llvm-project/libcxx/include/__memory/construct_at.h:88:5 (clickhouse+0x19e8bf1d)
E               #5 void std::__1::allocator_traits<std::__1::allocator<DB::KeeperStorageSnapshot>>::destroy[abi:v15000]<DB::KeeperStorageSnapshot, void, void>(std::__1::allocator<DB::KeeperStorageSnapshot>&, DB::KeeperStorageSnapshot*) build_docker/./contrib/llvm-project/libcxx/include/__memory/allocator_traits.h:317:9 (clickhouse+0x19e8bf1d)
E               #6 std::__1::__shared_ptr_emplace<DB::KeeperStorageSnapshot, std::__1::allocator<DB::KeeperStorageSnapshot>>::__on_zero_shared() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:309:9 (clickhouse+0x19e8bf1d)
E               #7 std::__1::__shared_count::__release_shared[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:174:9 (clickhouse+0x19db36ad) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #8 std::__1::__shared_weak_count::__release_shared[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:215:27 (clickhouse+0x19db36ad)
E               #9 std::__1::shared_ptr<DB::KeeperStorageSnapshot>::~shared_ptr[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__memory/shared_ptr.h:702:23 (clickhouse+0x19db36ad)
E               #10 DB::CreateSnapshotTask::~CreateSnapshotTask() build_docker/./src/Coordination/KeeperSnapshotManager.h:180:8 (clickhouse+0x19db36ad)
E               #11 DB::KeeperDispatcher::snapshotThread() build_docker/./src/Coordination/KeeperDispatcher.cpp:345:5 (clickhouse+0x19db36ad)
E               #12 DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2::operator()() const build_docker/./src/Coordination/KeeperDispatcher.cpp:428:53 (clickhouse+0x19dc01aa) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #13 decltype(std::declval<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&>()()) std::__1::__invoke[abi:v15000]<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x19dc01aa)
E               #14 decltype(auto) std::__1::__apply_tuple_impl[abi:v15000]<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&, std::__1::tuple<>&>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&, std::__1::tuple<>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/tuple:1789:1 (clickhouse+0x19dc01aa)
E               #15 decltype(auto) std::__1::apply[abi:v15000]<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&, std::__1::tuple<>&>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&, std::__1::tuple<>&) build_docker/./contrib/llvm-project/libcxx/include/tuple:1798:1 (clickhouse+0x19dc01aa)
E               #16 ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&)::'lambda'()::operator()() build_docker/./src/Common/ThreadPool.h:223:13 (clickhouse+0x19dc01aa)
E               #17 decltype(std::declval<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>()()) std::__1::__invoke[abi:v15000]<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&)::'lambda'()&>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0x19dc01aa)
E               #18 void std::__1::__invoke_void_return_wrapper<void, true>::__call<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&)::'lambda'()&>(ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&)::'lambda'()&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:479:9 (clickhouse+0x19dc01aa)
E               #19 std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&)::'lambda'(), void ()>::operator()[abi:v15000]() build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:235:12 (clickhouse+0x19dc01aa)
E               #20 void std::__1::__function::__policy_invoker<void ()>::__call_impl<std::__1::__function::__default_alloc_func<ThreadFromGlobalPoolImpl<true>::ThreadFromGlobalPoolImpl<DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2>(DB::KeeperDispatcher::initialize(Poco::Util::AbstractConfiguration const&, bool, bool, std::__1::shared_ptr<DB::Macros const> const&)::$_2&&)::'lambda'(), void ()>>(std::__1::__function::__policy_storage const*) build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:716:16 (clickhouse+0x19dc01aa)
E               #21 std::__1::__function::__policy_func<void ()>::operator()[abi:v15000]() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:848:16 (clickhouse+0xf1b6b4e) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #22 std::__1::function<void ()>::operator()() const build_docker/./contrib/llvm-project/libcxx/include/__functional/function.h:1187:12 (clickhouse+0xf1b6b4e)
E               #23 ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) build_docker/./src/Common/ThreadPool.cpp:455:13 (clickhouse+0xf1b6b4e)
E               #24 void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()::operator()() const build_docker/./src/Common/ThreadPool.cpp:217:73 (clickhouse+0xf1bca11) (BuildId: 403a4b24e6681e676dc250396280e21263dd56eb)
E               #25 decltype(std::declval<void>()()) std::__1::__invoke[abi:v15000]<void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(void&&) build_docker/./contrib/llvm-project/libcxx/include/__functional/invoke.h:394:23 (clickhouse+0xf1bca11)
E               #26 void std::__1::__thread_execute[abi:v15000]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>(std::__1::tuple<void, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>&, std::__1::__tuple_indices<>) build_docker/./contrib/llvm-project/libcxx/include/thread:284:5 (clickhouse+0xf1bca11)
E               #27 void* std::__1::__thread_proxy[abi:v15000]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, void ThreadPoolImpl<std::__1::thread>::scheduleImpl<void>(std::__1::function<void ()>, Priority, std::__1::optional<unsigned long>, bool)::'lambda0'()>>(void*) build_docker/./contrib/llvm-project/libcxx/include/thread:295:5 (clickhouse+0xf1bca11)
```

flag `snapshot_mode` is unprotected.

cc @antonio2368

--- a/src/Coordination/KeeperDispatcher.cpp
+++ b/src/Coordination/KeeperDispatcher.cpp
@@ -325,12 +325,12 @@ void KeeperDispatcher::snapshotThread()
         if (!snapshots_queue.pop(task))
             break;
 
-        if (shutdown_called)
-            break;
-
         try
         {
-            auto snapshot_file_info = task.create_snapshot(std::move(task.snapshot));
+            auto snapshot_file_info = task.create_snapshot(std::move(task.snapshot), /*execute_only_cleanup=*/shutdown_called);
+
+            if (shutdown_called)
+                break;
 
             if (snapshot_file_info.path.empty())
                 continue;

--- a/src/Coordination/KeeperSnapshotManager.h
+++ b/src/Coordination/KeeperSnapshotManager.h
@@ -98,8 +98,7 @@ struct SnapshotFileInfo
 };
 
 using KeeperStorageSnapshotPtr = std::shared_ptr<KeeperStorageSnapshot>;
-using CreateSnapshotCallback = std::function<SnapshotFileInfo(KeeperStorageSnapshotPtr &&)>;
-
+using CreateSnapshotCallback = std::function<SnapshotFileInfo(KeeperStorageSnapshotPtr &&, bool)>;
 
 using SnapshotMetaAndStorage = std::pair<SnapshotMetadataPtr, KeeperStoragePtr>;
 

--- a/src/Coordination/KeeperStateMachine.cpp
+++ b/src/Coordination/KeeperStateMachine.cpp
@@ -564,63 +564,65 @@ void KeeperStateMachine::create_snapshot(nuraft::snapshot & s, nuraft::async_res
     }
 
     /// create snapshot task for background execution (in snapshot thread)
-    snapshot_task.create_snapshot = [this, when_done](KeeperStorageSnapshotPtr && snapshot)
+    snapshot_task.create_snapshot = [this, when_done](KeeperStorageSnapshotPtr && snapshot, bool execute_only_cleanup)
     {
         nuraft::ptr<std::exception> exception(nullptr);
         bool ret = true;
-        try
+        if (!execute_only_cleanup)
         {
-            { /// Read storage data without locks and create snapshot
-                std::lock_guard lock(snapshots_lock);
+            try
+            {
+                { /// Read storage data without locks and create snapshot
+                    std::lock_guard lock(snapshots_lock);
 
-                if (latest_snapshot_meta && snapshot->snapshot_meta->get_last_log_idx() <= latest_snapshot_meta->get_last_log_idx())
-                {
-                    LOG_INFO(
-                        log,
-                        "Will not create a snapshot with last log idx {} because a snapshot with bigger last log idx ({}) is already "
-                        "created",
-                        snapshot->snapshot_meta->get_last_log_idx(),
-                        latest_snapshot_meta->get_last_log_idx());
-                }
-                else
-                {
-                    latest_snapshot_meta = snapshot->snapshot_meta;
-                    /// we rely on the fact that the snapshot disk cannot be changed during runtime
-                    if (isLocalDisk(*keeper_context->getLatestSnapshotDisk()))
+                    if (latest_snapshot_meta && snapshot->snapshot_meta->get_last_log_idx() <= latest_snapshot_meta->get_last_log_idx())
                     {
-                        auto snapshot_info = snapshot_manager.serializeSnapshotToDisk(*snapshot);
-                        latest_snapshot_info = std::move(snapshot_info);
-                        latest_snapshot_buf = nullptr;
+                        LOG_INFO(
+                            log,
+                            "Will not create a snapshot with last log idx {} because a snapshot with bigger last log idx ({}) is already "
+                            "created",
+                            snapshot->snapshot_meta->get_last_log_idx(),
+                            latest_snapshot_meta->get_last_log_idx());
                     }
                     else
                     {
-                        auto snapshot_buf = snapshot_manager.serializeSnapshotToBuffer(*snapshot);
-                        auto snapshot_info = snapshot_manager.serializeSnapshotBufferToDisk(*snapshot_buf, snapshot->snapshot_meta->get_last_log_idx());
-                        latest_snapshot_info = std::move(snapshot_info);
-                        latest_snapshot_buf = std::move(snapshot_buf);
-                    }
+                        latest_snapshot_meta = snapshot->snapshot_meta;
+                        /// we rely on the fact that the snapshot disk cannot be changed during runtime
+                        if (isLocalDisk(*keeper_context->getLatestSnapshotDisk()))
+                        {
+                            auto snapshot_info = snapshot_manager.serializeSnapshotToDisk(*snapshot);
+                            latest_snapshot_info = std::move(snapshot_info);
+                            latest_snapshot_buf = nullptr;
+                        }
+                        else
+                        {
+                            auto snapshot_buf = snapshot_manager.serializeSnapshotToBuffer(*snapshot);
+                            auto snapshot_info = snapshot_manager.serializeSnapshotBufferToDisk(*snapshot_buf, snapshot->snapshot_meta->get_last_log_idx());
+                            latest_snapshot_info = std::move(snapshot_info);
+                            latest_snapshot_buf = std::move(snapshot_buf);
+                        }
 
-                    ProfileEvents::increment(ProfileEvents::KeeperSnapshotCreations);
-                    LOG_DEBUG(log, "Created persistent snapshot {} with path {}", latest_snapshot_meta->get_last_log_idx(), latest_snapshot_info.path);
+                        ProfileEvents::increment(ProfileEvents::KeeperSnapshotCreations);
+                        LOG_DEBUG(log, "Created persistent snapshot {} with path {}", latest_snapshot_meta->get_last_log_idx(), latest_snapshot_info.path);
+                    }
                 }
             }
-
+            catch (...)
             {
-                /// Destroy snapshot with lock
-                std::lock_guard lock(storage_and_responses_lock);
-                LOG_TRACE(log, "Clearing garbage after snapshot");
-                /// Turn off "snapshot mode" and clear outdate part of storage state
-                storage->clearGarbageAfterSnapshot();
-                LOG_TRACE(log, "Cleared garbage after snapshot");
-                snapshot.reset();
+                ProfileEvents::increment(ProfileEvents::KeeperSnapshotCreationsFailed);
+                LOG_TRACE(log, "Exception happened during snapshot");
+                tryLogCurrentException(log);
+                ret = false;
             }
         }
-        catch (...)
         {
-            ProfileEvents::increment(ProfileEvents::KeeperSnapshotCreationsFailed);
-            LOG_TRACE(log, "Exception happened during snapshot");
-            tryLogCurrentException(log);
-            ret = false;
+            /// Destroy snapshot with lock
+            std::lock_guard lock(storage_and_responses_lock);
+            LOG_TRACE(log, "Clearing garbage after snapshot");
+            /// Turn off "snapshot mode" and clear outdate part of storage state
+            storage->clearGarbageAfterSnapshot();
+            LOG_TRACE(log, "Cleared garbage after snapshot");
+            snapshot.reset();
         }
 
         when_done(ret, exception);
@@ -628,11 +630,10 @@ void KeeperStateMachine::create_snapshot(nuraft::snapshot & s, nuraft::async_res
         return ret ? latest_snapshot_info : SnapshotFileInfo{};
     };
 
-
     if (keeper_context->getServerState() == KeeperContext::Phase::SHUTDOWN)
     {
         LOG_INFO(log, "Creating a snapshot during shutdown because 'create_snapshot_on_exit' is enabled.");
-        auto snapshot_file_info = snapshot_task.create_snapshot(std::move(snapshot_task.snapshot));
+        auto snapshot_file_info = snapshot_task.create_snapshot(std::move(snapshot_task.snapshot), /*execute_only_cleanup=*/false);
 
         if (!snapshot_file_info.path.empty() && snapshot_manager_s3)
         {

--- a/src/Coordination/tests/gtest_coordination.cpp
+++ b/src/Coordination/tests/gtest_coordination.cpp
@@ -1818,7 +1818,7 @@ void testLogAndStateMachine(
             bool pop_result = snapshots_queue.pop(snapshot_task);
             EXPECT_TRUE(pop_result);
 
-            snapshot_task.create_snapshot(std::move(snapshot_task.snapshot));
+            snapshot_task.create_snapshot(std::move(snapshot_task.snapshot), /*execute_only_cleanup=*/false);
         }
 
         if (snapshot_created && changelog.size() > settings->reserved_log_items)

